File: flutter/lib/models/compartilhamento.dart
class Compartilhamento {
  final int id;
  final int corretorId;
  final String entityType; // 'empreendimento' ou 'unidade'
  final int entityId;
  final String linkUnico;
  final String urlCompleta;
  final String? nomeCliente;
  final String? anotacao;
  final bool receberNotificacao;
  final bool mostrarEspelhoVendas;
  final bool mostrarEndereco;
  final bool compartilharDescricao;
  final int totalVisualizacoes;
  final DateTime? ultimaVisualizacaoAt;
  final bool ativo;
  final bool isExpirado;
  final DateTime? expiraEm;
  final DateTime createdAt;
  final DateTime updatedAt;
  final CompartilhamentoEntity? entity;

  Compartilhamento({
    required this.id,
    required this.corretorId,
    required this.entityType,
    required this.entityId,
    required this.linkUnico,
    required this.urlCompleta,
    this.nomeCliente,
    this.anotacao,
    required this.receberNotificacao,
    required this.mostrarEspelhoVendas,
    required this.mostrarEndereco,
    required this.compartilharDescricao,
    required this.totalVisualizacoes,
    this.ultimaVisualizacaoAt,
    required this.ativo,
    required this.isExpirado,
    this.expiraEm,
    required this.createdAt,
    required this.updatedAt,
    this.entity,
  });

  factory Compartilhamento.fromJson(Map<String, dynamic> json) {
    return Compartilhamento(
      id: json['id'] as int,
      corretorId: json['corretor_id'] as int,
      entityType: json['entity_type'] as String,
      entityId: json['entity_id'] as int,
      linkUnico: json['link_unico'] as String,
      urlCompleta: json['url_completa'] as String,
      nomeCliente: json['nome_cliente'] as String?,
      anotacao: json['anotacao'] as String?,
      receberNotificacao: json['receber_notificacao'] as bool? ?? false,
      mostrarEspelhoVendas: json['mostrar_espelho_vendas'] as bool? ?? false,
      mostrarEndereco: json['mostrar_endereco'] as bool? ?? true,
      compartilharDescricao: json['compartilhar_descricao'] as bool? ?? true,
      totalVisualizacoes: json['total_visualizacoes'] as int? ?? 0,
      ultimaVisualizacaoAt: json['ultima_visualizacao_at'] != null
          ? DateTime.parse(json['ultima_visualizacao_at'] as String)
          : null,
      ativo: json['ativo'] as bool? ?? true,
      isExpirado: json['is_expirado'] as bool? ?? false,
      expiraEm: json['expira_em'] != null
          ? DateTime.parse(json['expira_em'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      entity: json['entity'] != null
          ? CompartilhamentoEntity.fromJson(json['entity'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'corretor_id': corretorId,
      'entity_type': entityType,
      'entity_id': entityId,
      'link_unico': linkUnico,
      'url_completa': urlCompleta,
      'nome_cliente': nomeCliente,
      'anotacao': anotacao,
      'receber_notificacao': receberNotificacao,
      'mostrar_espelho_vendas': mostrarEspelhoVendas,
      'mostrar_endereco': mostrarEndereco,
      'compartilhar_descricao': compartilharDescricao,
      'total_visualizacoes': totalVisualizacoes,
      'ultima_visualizacao_at': ultimaVisualizacaoAt?.toIso8601String(),
      'ativo': ativo,
      'is_expirado': isExpirado,
      'expira_em': expiraEm?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'entity': entity?.toJson(),
    };
  }

  /// Retorna o nome de exibição da entidade
  String get entityNome => entity?.nome ?? 'N/A';

  /// Retorna se é um empreendimento
  bool get isEmpreendimento => entityType == 'empreendimento';

  /// Retorna se é uma unidade
  bool get isUnidade => entityType == 'unidade';

  /// Retorna uma descrição formatada do tipo
  String get tipoFormatado => isEmpreendimento ? 'Empreendimento' : 'Unidade';
}

class CompartilhamentoEntity {
  final int id;
  final String nome;

  CompartilhamentoEntity({
    required this.id,
    required this.nome,
  });

  factory CompartilhamentoEntity.fromJson(Map<String, dynamic> json) {
    return CompartilhamentoEntity(
      id: json['id'] as int,
      nome: json['nome'] as String? ?? 'N/A',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nome': nome,
    };
  }
}

class CompartilhamentoEstatisticas {
  final int totalVisualizacoes;
  final int totalAcessos;
  final int acessosUnicos;
  final DateTime? ultimaVisualizacao;
  final DateTime criadoEm;
  final List<AcessoPorData> acessosPorData;

  CompartilhamentoEstatisticas({
    required this.totalVisualizacoes,
    required this.totalAcessos,
    required this.acessosUnicos,
    this.ultimaVisualizacao,
    required this.criadoEm,
    required this.acessosPorData,
  });

  factory CompartilhamentoEstatisticas.fromJson(Map<String, dynamic> json) {
    return CompartilhamentoEstatisticas(
      totalVisualizacoes: json['total_visualizacoes'] as int? ?? 0,
      totalAcessos: json['total_acessos'] as int? ?? 0,
      acessosUnicos: json['acessos_unicos'] as int? ?? 0,
      ultimaVisualizacao: json['ultima_visualizacao'] != null
          ? DateTime.parse(json['ultima_visualizacao'] as String)
          : null,
      criadoEm: DateTime.parse(json['criado_em'] as String),
      acessosPorData: (json['acessos_por_data'] as List<dynamic>?)
              ?.map((e) => AcessoPorData.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class AcessoPorData {
  final String data;
  final int total;

  AcessoPorData({
    required this.data,
    required this.total,
  });

  factory AcessoPorData.fromJson(Map<String, dynamic> json) {
    return AcessoPorData(
      data: json['data'] as String,
      total: json['total'] as int,
    );
  }
}

class CriarCompartilhamentoRequest {
  final String entityType;
  final int entityId;
  final String? nomeCliente;
  final String? anotacao;
  final bool receberNotificacao;
  final bool mostrarEspelhoVendas;
  final bool mostrarEndereco;
  final bool compartilharDescricao;
  final DateTime? expiraEm;

  CriarCompartilhamentoRequest({
    required this.entityType,
    required this.entityId,
    this.nomeCliente,
    this.anotacao,
    this.receberNotificacao = false,
    this.mostrarEspelhoVendas = false,
    this.mostrarEndereco = true,
    this.compartilharDescricao = true,
    this.expiraEm,
  });

  Map<String, dynamic> toJson() {
    return {
      'entity_type': entityType,
      'entity_id': entityId,
      if (nomeCliente != null && nomeCliente!.isNotEmpty) 'nome_cliente': nomeCliente,
      if (anotacao != null && anotacao!.isNotEmpty) 'anotacao': anotacao,
      'receber_notificacao': receberNotificacao,
      'mostrar_espelho_vendas': mostrarEspelhoVendas,
      'mostrar_endereco': mostrarEndereco,
      'compartilhar_descricao': compartilharDescricao,
      if (expiraEm != null) 'expira_em': expiraEm!.toIso8601String(),
    };
  }
}

File: app/Models/Compartilhamento.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Str;

class Compartilhamento extends Model
{
    use HasFactory;

    protected $table = 'compartilhamentos';

    protected $fillable = [
        'corretor_id',
        'entity_type',
        'entity_id',
        'link_unico',
        'nome_cliente',
        'anotacao',
        'receber_notificacao',
        'mostrar_espelho_vendas',
        'mostrar_endereco',
        'compartilhar_descricao',
        'total_visualizacoes',
        'ultima_visualizacao_at',
        'ativo',
        'expira_em',
    ];

    protected $casts = [
        'receber_notificacao' => 'boolean',
        'mostrar_espelho_vendas' => 'boolean',
        'mostrar_endereco' => 'boolean',
        'compartilhar_descricao' => 'boolean',
        'total_visualizacoes' => 'integer',
        'ativo' => 'boolean',
        'ultima_visualizacao_at' => 'datetime',
        'expira_em' => 'datetime',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Relacionamento com corretor
     */
    public function corretor()
    {
        return $this->belongsTo(Corretor::class);
    }

    /**
     * Relacionamento com empreendimento (polimórfico)
     */
    public function entity()
    {
        return $this->morphTo('entity', 'entity_type', 'entity_id');
    }

    /**
     * Relacionamento com empreendimento
     */
    public function empreendimento()
    {
        return $this->belongsTo(Empreendimento::class, 'entity_id');
    }

    /**
     * Relacionamento com unidade
     */
    public function unidade()
    {
        return $this->belongsTo(EmpreendimentoUnidade::class, 'entity_id');
    }

    /**
     * Relacionamento com acessos
     */
    public function acessos()
    {
        return $this->hasMany(AcessoCompartilhamento::class);
    }

    /**
     * Relacionamento com click tracking
     */
    public function clickTrackings()
    {
        return $this->hasMany(ClickTracking::class);
    }

    /**
     * Gera um link único para o compartilhamento
     */
    public function gerarLinkUnico(): string
    {
        do {
            $linkUnico = Str::random(32);
        } while (self::where('link_unico', $linkUnico)->exists());

        return $linkUnico;
    }

    /**
     * Incrementa o contador de visualizações
     */
    public function incrementarVisualizacao(): void
    {
        $this->increment('total_visualizacoes');
        $this->update(['ultima_visualizacao_at' => now()]);
    }

    /**
     * Retorna a URL completa do compartilhamento
     */
    public function getUrlCompletaAttribute(): string
    {
        $baseUrl = config('app.url', 'https://backend.valeincorp.com.br');
        return "{$baseUrl}/share/{$this->link_unico}";
    }

    /**
     * Verifica se o link está expirado
     */
    public function isExpirado(): bool
    {
        if (!$this->expira_em) {
            return false;
        }

        return now()->isAfter($this->expira_em);
    }

    /**
     * Verifica se o link está ativo e não expirado
     */
    public function isAtivo(): bool
    {
        return $this->ativo && !$this->isExpirado();
    }

    /**
     * Scope para filtrar compartilhamentos ativos
     */
    public function scopeAtivos($query)
    {
        return $query->where('ativo', true)
            ->where(function ($q) {
                $q->whereNull('expira_em')
                  ->orWhere('expira_em', '>', now());
            });
    }

    /**
     * Scope para filtrar por corretor
     */
    public function scopePorCorretor($query, $corretorId)
    {
        return $query->where('corretor_id', $corretorId);
    }

    /**
     * Scope para filtrar por tipo de entidade
     */
    public function scopePorTipoEntidade($query, $entityType)
    {
        return $query->where('entity_type', $entityType);
    }

    /**
     * Boot method para gerar link único automaticamente
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($compartilhamento) {
            if (empty($compartilhamento->link_unico)) {
                $compartilhamento->link_unico = $compartilhamento->gerarLinkUnico();
            }
        });
    }
}
